man/man1/lgt2pdf.1
.TH lgt2pdf 1 "December 1, 2023" "Logtalk 3.73.0" "Logtalk Documentation"

.SH NAME
lgt2pdf \- Logtalk script for converting XML documenting files into PDF files

.SH SYNOPSIS
.B lgt2pdf
[\fB-f \fIformat\fR]
[\fB-d \fIdirectory\fR]
[\fB-p \fIprocessor\fR]
.br
.B lgt2pdf
.B \-h

.SH DESCRIPTION
\fBlgt2pdf\fR is a Logtalk script for converting all XML documenting files in the current directory into PDF files.

.SH OPTIONS
.TP
.B \-f
.I format
Paper format (either \fIa4\fR or \fIus\fR).
.TP
.B \-d
.I directory
Output directory for the generated PDF files.
.TP
.B \-p
.I processor
XSL-FO processor.
.TP
.B \-h
Prints a summary of the script options.

.SH REQUIREMENTS
Requires a compatible XSL-FO processor: \fIfop\fR (default), \fIfop2\fR, \fIxep\fR, or \fIxinc\fR.
The \fIfop2\fR processor requires generating the XML documenting files using the option \fIxml_spec(xsd)\fR or \fIxml_spec_reference(standalone)\fR.

.SH ENVIRONMENT
.TP
.B LOGTALKHOME
Specifies the location of the Logtalk installation.
.TP
.B LOGTALKUSER
Specifies the location of the Logtalk user directory.

.SH FILES
.TP
.BI $LOGTALKHOME/tools/lgtdoc/xml/logtalk_entity.dtd
The Logtalk entity DTD file.
.TP
.BI $LOGTALKHOME/tools/lgtdoc/xml/logtalk_index.dtd
The Logtalk index DTD file.
.TP
.BI $LOGTALKUSER/tools/lgtdoc/xml/custom.ent
The Logtalk custom entities file.
.TP
.BI $LOGTALKHOME/tools/lgtdoc/xml/logtalk_entity.xsd
The Logtalk entity XML Schema file.
.TP
.BI $LOGTALKHOME/tools/lgtdoc/xml/logtalk_index.xsd
The Logtalk index XML Schema file.
.TP
.BI $LOGTALKUSER/tools/lgtdoc/xml/lgtpdfa4.xsl
The default XSLT file for generating PDF A4 files.
.TP
.BI $LOGTALKUSER/tools/lgtdoc/xml/logtalk_entity_to_pdf.xsl
The Logtalk entity XSLT file.
.TP
.BI $LOGTALKUSER/tools/lgtdoc/xml/logtalk_index_to_pdf.xsl
The Logtalk index XSLT file.

.SH "SEE ALSO"
The Logtalk website at https://logtalk.org/
.PP
The Logtalk User and Reference Manuals at \fB$LOGTALKHOME/manuals/index.html\fR
.PP
logtalk_user_setup(1),\ lgt2html(1),\ lgt2xml(1),\ lgt2md(1),\ lgt2txt(1)

.SH AUTHOR
Paulo Moura <pmoura@logtalk.org>

.SH COPYRIGHT
Copyright (c) 1998-2021 Paulo Moura.


RELEASE_NOTES.md
________________________________________________________________________

This file is part of Logtalk <https://logtalk.org/>  
SPDX-FileCopyrightText: 1998-2023 Paulo Moura <pmoura@logtalk.org>  
SPDX-License-Identifier: Apache-2.0

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
________________________________________________________________________


RELEASE NOTES
=============


3.73.0 - December ??, 2023
==========================

Logtalk compiler and runtime
----------------------------

* ADDED: Linter warnings for deprecated arithmetic functions.

* ADDED: Linter warning for comparing numbers using unification.

* ADDED: Linter warnings for Quintus Prolog predicate versions of arithmetic
functions.

* ADDED: Additional legacy built-in predicates for linter deprecated warnings.

* ADDED: Support for using backend-declared deprecated built-in predicates in
linter warnings.

* IMPROVED: When checking the availability of predicates in `user` for `uses/2`
and `use_module/2` directives, also check for an `imported_from/1` predicate
property.

* IMPROVED: Avoid false positive linter warnings about non-terminals called as
predicates when the caller is a phrase-like predicate declared in the backend
adapter file. Thanks to Yurii Rashkovskii for the use-case report.

* IMPROVED: Compiler reporting of term-expansion errors to write the source
term.

* FIXED: Unknown and undefined predicate call warnings when the calls occur in
an included file to report the actual location instead of the main file.

* FIXED: Only print linter warnings related to grammar rules when the flag
`grammar_rules` is set to `warning`.

Prolog adapter and integration files
------------------------------------

* ADDED: Support for declaring phrase-like predicates that call non-terminals.
This helps preventing false positive linter warnings of non-terminals called
as predicates.

* ADDED: Support for declaring deprecated built-in predicates (with or without
advised alternatives) in the adapter files.

* IMPROVED: ECLiPSe definition of the `'$lgt_current_module_predicate'/2`
internal predicate.

* UPDATED: The CxProlog adapter file to require version 0.98.2 or later.

* UPDATED: The LVM adapter file to prevent a clash with the `term_hash/4`
built-in predicate introduced in an upcoming version.

* UPDATED: The SWI-Prolog adapter file to load the `prolog_evaluable` library
when available (provides the `evaluable_property/2` predicate used by the
linter).

Documentation
-------------

* IMPROVED: Handbook grammar section now uses W3C-style EBNF syntax compatible
with the [Railroad Diagram Generator](https://www.bottlecaps.de/rr/ui). Several
typos and omissions are also fixed. Thanks to Domingo Alvarez Duarte for his
contribution and bug report.

* IMPROVED: Documentation of the `wrapper` tool.

* UPDATED: The manual installation instructions on generating documentation
human-readable formats.

* UPDATED: The man page for the `lgt2pdf` shell script.

Library
-------

* REMOVED: The `tsv` object from the `csv` library.

* ADDED: Experimental `mutations` library for generating random mutations of
terms of selected types. Intended for eventual fuzz testing support.

* ADDED: New `tsv` library for reading and writing TSV files.

* ADDED: Library predicates `list::occurrences/2-3` to count the number of
occurrences of list elements.

* ADDED: Predicates `select/4`, `swap/2`, and `swap_consecutive/2` to the
`random` library for returning a random member of a list, replacing it by
a new element and returning the resulting list, and for swapping two randomly
chosen elements of a list.

* ADDED: Non-terminals `lazy_without//2`, `greedy_without//2`, and `rest//1`
to the `sequence_grammars` library object.

* IMPROVED: Performance of the `reader` library predicates that read a file
or a stream to a list of chars, codes, or bytes.

* IMPROVED: Performance of the `term_io` library predicates.

* UPDATED: The `csv` library to also recognize `.tab` file extension for TSV
files.

* FIXED: The documentation of the `pairs::keys_values/3` library predicate.

* FIXED: The `os` library tests calls to the `shell/1` predicate from auxiliary
predicates.

* FIXED: The `listing` library category implementation of the `listing/1`
predicate.

Tools
--------

* UPDATED: The `lgtunit` tool to also print the tests runtime in seconds.

* UPDATED: The `tutor` tool for the new arithmetic function linter warnings
and the new linter warning for comparing numbers using unification.

* FIXED: The `debugger` tool checking for a deterministic goal for CxProlog.

* FIXED: The `lgtunit` tool definition of the `deterministic/1-2` predicates
for CxProlog.

* FIXED: The `logtalk_tester` scripts to detect and report as broken test sets
where the tests failed to run due to an unexpected loading error likely caused
by a backend bug.

* FIXED: The `lgt2html.ps1`, `lgt2md.ps1`, `lgt2rst.ps1`, and `lgt2txt.ps1`
PowerShell scripts to release the lock on the XML files after converting them
to reStructuredText files.

* FIXED: Error in the `wrapper` tool when printing missing directives advice.

Examples
--------

* ADDED: New `scratchcards` example, providing a solution for the Advent of
Code 2023 Day 4 problem.

* ADDED: New `trebuchet` example, providing a solution for the Advent of Code
2023 Day 1 problem.

* ADDED: Tests to the `poem`, `profiling`, and `self_vs_this` examples.

* UPDATED: The `errors` example for the new arithmetic function linter
warnings.


tools/lgtdoc/xml/lgt2pdf.ps1
#############################################################################
## 
##   XML documenting files to PDF conversion script 
##   Last updated on March 15, 2023
## 
##   This file is part of Logtalk <https://logtalk.org/>  
##   Copyright 2022 Hans N. Beck and Paulo Moura <pmoura@logtalk.org>
##   SPDX-License-Identifier: Apache-2.0
##   
##   Licensed under the Apache License, Version 2.0 (the "License");
##   you may not use this file except in compliance with the License.
##   You may obtain a copy of the License at
##   
##       http://www.apache.org/licenses/LICENSE-2.0
##   
##   Unless required by applicable law or agreed to in writing, software
##   distributed under the License is distributed on an "AS IS" BASIS,
##   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
##   See the License for the specific language governing permissions and
##   limitations under the License.
## 
#############################################################################


#Requires -Version 7.3

[CmdletBinding()]
param(
	[Parameter()]
	[String]$f = "a4", 
	[String]$d = $pwd, 
	[String]$p = "fop", 
	[Switch]$v,
	[Switch]$h
)

function Write-Script-Version {
	$myFullName = $MyInvocation.ScriptName
	$myName = Split-Path -Path $myFullName -leaf -Resolve
	Write-Output ($myName + " 2.0")
}

function Get-Logtalkhome {
	if ($null -eq $env:LOGTALKHOME) 
	{
		Write-Output "The environment variable LOGTALKHOME should be defined first, pointing"
		Write-Output "to your Logtalk installation directory!"
		Write-Output "Trying the default locations for the Logtalk installation..."
		
		$DEFAULTPATHS = [string[]](
			"C:\Program Files (x86)\Logtalk",
			"C:\Program Files\Logtalk",
			"%LOCALAPPDATA%\Logtalk"
		)
		
		# Checking all possibilites
		foreach ($DEFAULTPATH in $DEFAULTPATHS) { 
			Write-Output ("Looking for: " + $DEFAULTPATH)
			if (Test-Path $DEFAULTPATH) {
				Write-Output ("... using Logtalk installation found at " + $DEFAULTPATH)
				$env:LOGTALKHOME = $DEFAULTPATH
				break
			}
		}
	}
	# At the end LOGTALKHOME was set already or now is set
}

function Get-Logtalkuser {
	if ($null -eq $env:LOGTALKUSER) {
		Write-Output "After the script completion, you must set the environment variable"
		Write-Output "LOGTALKUSER pointing to %USERPROFILE%\Documents\Logtalk."
		$env:LOGTALKUSER = "%USERPROFILE%\Documents\Logtalk"
	}
	# At the end LOGTALKUSER was set already or now is set
}

function Write-Usage-Help() {
	$myFullName = $MyInvocation.ScriptName
	$myName = Split-Path -Path $myFullName -leaf -Resolve 

	Write-Output "This script converts all Logtalk XML documenting files in the"
	Write-Output "current directory to PDF files"
	Write-Output ""
	Write-Output "Usage:"
	Write-Output ("  " + $myName + " [-f format] [-d directory] [-p processor]")
	Write-Output ("  " + $myName + " -v")
	Write-Output ("  " + $myName + " -h")
	Write-Output ""
	Write-Output "Optional arguments:"
	Write-Output ("  -f output file format (either a4 or us; default is " + $f + ")")
	Write-Output ("  -d output directory for the generated files (default is " + $d + ")")
	Write-Output ("  -p XSL-FO processor (either fop, fop2, xep, or xinc; default is " + $d + ")")
	Write-Output "  -v print version"
	Write-Output "  -h help"
	Write-Output ""
}

function Check-Parameters() {

	if ($f -ne "a4" -and $f -ne "us") {
		Write-Output ("Error! Unsupported output format: " + $f)
		Start-Sleep -Seconds 2
		Exit
	}

	if (-not(Test-Path $d)) { # cannot be ""
		Write-Output ("The " + $p + " output directory does not exist!")
		Start-Sleep -Seconds 2
		Exit
	}

	if ($p -ne "fop" -and $p -ne "fop2" -and $p -ne "xep" -and $p -ne "xinc") {
		Write-Output ("Error! Unsupported XSL-FO processor: " + $f)
		Start-Sleep -Seconds 2
		Exit
	}

	if ($v -eq $true) {
		Write-Script-Version
		Exit
	}

	if ($h -eq $true) {
		Write-Usage-Help
		Exit
	}

}

###################### here it starts ############################ 

Check-Parameters

Get-Logtalkhome

# Check for existence
if (Test-Path $env:LOGTALKHOME) {
	$output = "Found LOGTALKHOME at: " + $env:LOGTALKHOME
	Write-Output $output
} else {
	Write-Output "... unable to locate Logtalk installation directory!"
	Start-Sleep -Seconds 2
	Exit
}

Get-Logtalkuser

# Check for existence
if (Test-Path $env:LOGTALKUSER) {
	if (!(Test-Path $env:LOGTALKUSER/VERSION.txt)) {
		Write-Output "Cannot find version information in the Logtalk user directory at %LOGTALKUSER%!"
		Write-Output "Creating an up-to-date Logtalk user directory..."
		logtalk_user_setup
	} else {
		$system_version = Get-Content $env:LOGTALKHOME/VERSION.txt
		$user_version = Get-Content $env:LOGTALKUSER/VERSION.txt
		if ($user_version -lt $system_version) {
			Write-Output "Logtalk user directory at %LOGTALKUSER% is outdated: "
			Write-Output "    $user_version < $system_version"
			Write-Output "Creating an up-to-date Logtalk user directory..."
			logtalk_user_setup
		}
	}
} else {
	Write-Output "Cannot find %LOGTALKUSER% directory! Creating a new Logtalk user directory"
	Write-Output "by running the logtalk_user_setup shell script:"
	logtalk_user_setup
}

if ($f -eq "a4") {
	$entity_xslt =$env:LOGTALKUSER + "\tools\lgtdoc\xml\logtalk_entity_to_pdf_a4.xsl"
} else {
	$entity_xslt  =$env:LOGTALKUSER + "\tools\lgtdoc\xml\logtalk_entity_to_pdf_us.xsl"
}

if (!(Test-Path "logtalk_entity.dtd")) {
	Copy-Item -Path $env:LOGTALKHOME\tools\lgtdoc\xml\logtalk_entity.dtd -Destination .
}

if (!(Test-Path "custom.ent")) {
	Copy-Item -Path $env:LOGTALKUSER\tools\lgtdoc\xml\custom.ent -Destination .
}

if (!(Test-Path "logtalk_entity.xsd")) {
	Copy-Item -Path $env:LOGTALKHOME\tools\lgtdoc\xml\logtalk_entity.xsd -Destination .
}


if (Select-String -Path .\*.xml -Pattern '<logtalk' -CaseSensitive -SimpleMatch -Quiet) {
	Write-Output "Converting XML files to PDF files..."

	Get-ChildItem -Path .\*.xml |
	Foreach-Object {
		if (Select-String -Path $_ -Pattern '<logtalk_entity' -CaseSensitive -SimpleMatch -Quiet) {
			Write-Output ("  converting " + $_.Name)
			$file = Join-Path $pwd $_.Name
			$pdf = Join-Path $d ($_.BaseName + ".pdf")
			if ( $p -eq  "xinc") {
				 xinc -xml "$file" -xsl "$entity_xslt" -pdf "$pdf" > $null
			} else {
				 & $p -q -xml "$file" -xsl "$entity_xslt" -pdf "$pdf" > $null			
			}
		}
	}
	Write-Output "conversion done"
	Write-Output ""
} else {
	Write-Output ""
	Write-Output "No XML files exist in the current directory!"
	Write-Output ""
}

if ($pwd -ne (Join-Path $env:LOGTALKHOME xml)) {
	Remove-Item -Path .\logtalk_entity.dtd
	Remove-Item -Path .\logtalk_entity.xsd
	Remove-Item -Path .\custom.ent
}
